<a href="https://colab.research.google.com/github/Anastasiia-cn/Python_for_analytics_HW/blob/main/Kabanova_hw_3_2_%D0%9E%D1%81%D0%BD%D0%BE%D0%B2%D0%B8_%D1%80%D0%BE%D0%B1%D0%BE%D1%82%D0%B8_%D0%B7_%D0%B4%D0%B0%D0%BD%D0%B8%D0%BC%D0%B8_%D1%83_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import pandas as pd

df = pd.read_csv('/content/drive/MyDrive/Python_DA_course_HW/pandas/data/supermarket_sales.csv')

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [3]:
df.head(4)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4

In [4]:
df.tail(4)

Invoice ID Branch      City Customer type  Gender         Product line  \
996  303-96-2227      B  Mandalay        Normal  Female   Home and lifestyle   
997  727-02-1313      A    Yangon        Member    Male   Food and beverages   
998  347-56-2442      A    Yangon        Normal    Male   Home and lifestyle   
999  849-09-3807      A    Yangon        Member  Female  Fashion accessories   

     Unit price  Quantity  Tax 5%     Total       Date   Time  Payment  \
996       97.38        10  48.690  1022.490   3/2/2019  17:16  Ewallet   
997       31.84         1   1.592    33.432   2/9/2019  13:22     Cash   
998       65.82         1   3.291    69.111  2/22/2019  15:33     Cash   
999       88.34         7  30.919   649.299  2/18/2019  13:28     Cash   

       cogs  gross margin percentage  gross income  Rating  
996  973.80                 4.761905        48.690     4.4  
997   31.84                 4.761905         1.592     7.7  
998   65.82                 4.761905         3.291     4.1  
999  618.38                 4.761905        30.919     6.6

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [5]:
df[615:620]

Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   
619  641-96-3695      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   
619     Fashion accessories       43.46         6  13.0380  273.7980   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761905   
617  3/14/2019  18:10      Ewallet  439.55                 4.761905   
618  1/23/2019  11:22  Credit card  591.18                 4.761905   
619   2/7/2019  17:55      Ewallet  260.76                 4.761905   

     gross income  Rating  
615       17.4240     4.2  
616       21.7800     6.9  
617       21.9775     4.4  
618       29.5590     4.0  
619       13.0380     8.5

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [6]:
df.dtypes

Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [7]:
df['gross income'].describe(percentiles=[0.01, 0.25, 0.5, 0.75, 0.99])

count    1000.000000
mean       15.379369
std        11.708825
min         0.508500
1%          0.913735
25%         5.924875
50%        12.088000
75%        22.445250
99%        45.250750
max        49.650000
Name: gross income, dtype: float64

75%: 22.445250
\
Це означає, що 75% даних колонки 'gross income' мають значення менше 22.445250.


5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [8]:
df['City'].value_counts()

Yangon       340
Mandalay     332
Naypyitaw    328
Name: City, dtype: int64

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [9]:
df['City'].value_counts(normalize=True)

Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: City, dtype: float64

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [10]:
df[df['City'] == 'Mandalay'].head(10)

Invoice ID Branch      City Customer type  Gender  \
9   692-92-5582      B  Mandalay        Member  Female   
10  351-62-0822      B  Mandalay        Member  Female   
11  529-56-3974      B  Mandalay        Member    Male   
15  299-46-1805      B  Mandalay        Member  Female   
19  319-50-3348      B  Mandalay        Normal  Female   
21  371-85-5789      B  Mandalay        Normal    Male   
22  273-16-6619      B  Mandalay        Normal    Male   
26  649-29-6775      B  Mandalay        Normal    Male   
28  145-94-9061      B  Mandalay        Normal  Female   
30  871-79-8483      B  Mandalay        Normal    Male   

              Product line  Unit price  Quantity   Tax 5%     Total  \
9       Food and beverages       54.84         3   8.2260  172.7460   
10     Fashion accessories       14.48         4   2.8960   60.8160   
11  Electronic accessories       25.51         4   5.1020  107.1420   
15       Sports and travel       93.72         6  28.1160  590.4360   
19      Home and lifestyle       40.30         2   4.0300   84.6300   
21       Health and beauty       87.98         3  13.1970  277.1370   
22      Home and lifestyle       33.20         2   3.3200   69.7200   
26     Fashion accessories       33.52         1   1.6760   35.1960   
28      Food and beverages       88.36         5  22.0900  463.8900   
30     Fashion accessories       94.13         5  23.5325  494.1825   

         Date   Time      Payment    cogs  gross margin percentage  \
9   2/20/2019  13:27  Credit card  164.52                 4.761905   
10   2/6/2019  18:07      Ewallet   57.92                 4.761905   
11   3/9/2019  17:03         Cash  102.04                 4.761905   
15  1/15/2019  16:19         Cash  562.32                 4.761905   
19  3/11/2019  15:30      Ewallet   80.60                 4.761905   
21   3/5/2019  10:40      Ewallet  263.94                 4.761905   
22  3/15/2019  12:20  Credit card   66.40                 4.761905   
26   2/8/2019  15:31         Cash   33.52                 4.761905   
28  1/25/2019  19:48         Cash  441.80                 4.761905   
30  2/25/2019  19:39  Credit card  470.65                 4.761905   

    gross income  Rating  
9         8.2260     5.9  
10        2.8960     4.5  
11        5.1020     6.8  
15       28.1160     4.5  
19        4.0300     4.4  
21       13.1970     5.1  
22        3.3200     4.4  
26        1.6760     6.7  
28       22.0900     9.6  
30       23.5325     4.8

In [11]:
df[df['City'] == 'Yangon'].head(10)

Invoice ID Branch    City Customer type  Gender            Product line  \
0   750-67-8428      A  Yangon        Member  Female       Health and beauty   
2   631-41-3108      A  Yangon        Normal    Male      Home and lifestyle   
3   123-19-1176      A  Yangon        Member    Male       Health and beauty   
4   373-73-7910      A  Yangon        Normal    Male       Sports and travel   
6   355-53-5943      A  Yangon        Member  Female  Electronic accessories   
8   665-32-9167      A  Yangon        Member  Female       Health and beauty   
12  365-64-0515      A  Yangon        Normal  Female  Electronic accessories   
13  252-56-2699      A  Yangon        Normal    Male      Food and beverages   
14  829-34-3910      A  Yangon        Normal  Female       Health and beauty   
16  656-95-9349      A  Yangon        Member  Female       Health and beauty   

    Unit price  Quantity   Tax 5%     Total       Date   Time      Payment  \
0        74.69         7  26.1415  548.9715   1/5/2019  13:08      Ewallet   
2        46.33         7  16.2155  340.5255   3/3/2019  13:23  Credit card   
3        58.22         8  23.2880  489.0480  1/27/2019  20:33      Ewallet   
4        86.31         7  30.2085  634.3785   2/8/2019  10:37      Ewallet   
6        68.84         6  20.6520  433.6920  2/25/2019  14:36      Ewallet   
8        36.26         2   3.6260   76.1460  1/10/2019  17:15  Credit card   
12       46.95         5  11.7375  246.4875  2/12/2019  10:25      Ewallet   
13       43.19        10  21.5950  453.4950   2/7/2019  16:48      Ewallet   
14       71.38        10  35.6900  749.4900  3/29/2019  19:21         Cash   
16       68.93         7  24.1255  506.6355  3/11/2019  11:03  Credit card   

      cogs  gross margin percentage  gross income  Rating  
0   522.83                 4.761905       26.1415     9.1  
2   324.31                 4.761905       16.2155     7.4  
3   465.76                 4.761905       23.2880     8.4  
4   604.17                 4.761905       30.2085     5.3  
6   413.04                 4.761905       20.6520     5.8  
8    72.52                 4.761905        3.6260     7.2  
12  234.75                 4.761905       11.7375     7.1  
13  431.90                 4.761905       21.5950     8.2  
14  713.80                 4.761905       35.6900     5.7  
16  482.51                 4.761905       24.1255     4.6

In [12]:
yangon_gross_income_mean = df[df['City'] == 'Yangon']['gross income'].mean()
mandalay_gross_income_mean = df[df['City'] == 'Mandalay']['gross income'].mean()
print(f'Yangon gross income mean: {round(yangon_gross_income_mean, 2)}')
print(f'Mandalay gross income mean: {round(mandalay_gross_income_mean, 2)}')


Yangon gross income mean: 14.87
Mandalay gross income mean: 15.23


Гіпотеза 1: у Mandalay більше відвідувачів, які здійснюють покупки.
\
Гіпотеза 2: у Mandalay відвідувачі купують більше товарів.
\
Гіпотеза 3: у Mandalay відвідувачі купують дорожчі товари.
\
Перевірити гіпотези можна за рахунок обчислення загальної кількості відвідувачів, середньої кількості товарів у чеку та середнього значення Unit price у кожному місті.



8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [13]:
if len(df['gross margin percentage'].unique()) == 1:
  print('У всіх інвойсах однакова маржа')
else:
  print('Маржа відрізняється для різних інвойсів')

У всіх інвойсах однакова маржа


9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [15]:
df[df['City'] == 'Yangon']['Product line'].value_counts(normalize=True)


Home and lifestyle        0.191176
Electronic accessories    0.176471
Sports and travel         0.173529
Food and beverages        0.170588
Fashion accessories       0.150000
Health and beauty         0.138235
Name: Product line, dtype: float64

In [16]:
df[df['City'] == 'Mandalay']['Product line'].value_counts(normalize=True)

Fashion accessories       0.186747
Sports and travel         0.186747
Electronic accessories    0.165663
Health and beauty         0.159639
Food and beverages        0.150602
Home and lifestyle        0.150602
Name: Product line, dtype: float64